UnitySetup/UnitySetup.psm1

```diff
@@ -2132,3 +2132,565 @@ function Get-UnityLicense {
+            Start-Process "https://dev.azure.com/$($OrgName)/_usersSettings/tokens"
+        }
+
+        $GoodPAT = $false
+        while (-not $GoodPAT) {
+            $UserPAT = Read-Host -Prompt "Please enter your PAT for $($OrgName)"
+            if (Confirm-PAT "$($OrgName)" "$($ProjectName)" "$($FeedName)" "$($UserPAT.trim())") {
+                return [System.Convert]::ToBase64String([System.Text.Encoding]::UTF8.GetBytes(":" + $UserPAT.trim()))
+                $GoodPAT = $true
+            }
+            else {
+                Write-Host "Unable to validate PAT, please try again"
+            }
+        }
+    }
+
+    function New-PAT($PATName, $OrgName, $Scopes, $ExpireDays) {
+        $expireDate = (Get-Date).adddays($ExpireDays).ToString('yyyy-MM-ddTHH:mm:ss.fffZ')
+
+        $createPAT = 'y'
+
+        if (-not $env:ADO_BUILD_ENVIRONMENT) {
+            $answer = Read-Host "A Personal Access Token (PAT) will be created for you with the following details
+    Name: $PATName
+    Organization: $OrgName
+    Expiration: $expireDate
+Would you like to continue? (Default: $($createPAT))
+"
+            if (-not [string]::IsNullOrEmpty($answer)) {
+                $createPAT = $answer
+            }
+        }
+
+        if (($createPAT -like 'y') -or ($createPAT -like 'yes')) {
+        }
+        else {
+            return $null
+        }
+
+        if (-not (Get-Module -ListAvailable "Az.Accounts")) {
+            if (-not $isRunAsAdministrator) {
+                Write-Error "This script requires admin permissions to install a module for Azure Accounts (used to log you in and create PATs for you).
+             Please restart the script in an admin console, or run the script with the -ManualPAT option and supply your own PATs when prompted."
+            }
+            [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+            Install-Module -Name Az.Accounts -AllowClobber -Repository PSGallery -Scope CurrentUser -Force | Out-Null
+        }
+
+        if (-not (Get-Module -ListAvailable "Az.Accounts")) {
+            Write-Error "Unable to find the az.accounts module. Please check previous errors and/or restart Powershell and try again. If it's still not working, run with the `-ManualPAT` flag to go through the interactive flow of manually creating a PAT."
+            exit 1
+        }
+
+        if (-not $env:ADO_BUILD_ENVIRONMENT) {
+            $azaccount = $(Get-AzContext).Account
+
+            if ([string]::IsNullOrEmpty($azaccount) -or (-not $azaccount.Id.Contains("@microsoft.com"))) {
+                Write-Host "Connecting to Azure, please login if prompted"
+                Connect-AzAccount | Out-Null
+            }
+            $AZTokenRequest = Get-AzAccessToken -ResourceType Arm
+            $headers = @{Authorization = "Bearer $($AZTokenRequest.Token)" }
+        }
+        else {
+            $headers = @{Authorization = "Bearer $env:SYSTEM_ACCESSTOKEN" }
```

Comment:
I feel like we should not be adding ADO pipeline logic into this - can pipelines use the `ManualPAT` method (maybe we should rename that to `PersonalAccessToken`? That way it can function the same regardless of what pipeline setup users might have?

Reply:
Hmm I don't think there's a realistic way for a pipeline agent to use the ManualPAT approach since it opens a browser window and requires input.  I may need to leave this one in because I'm not certain the level of pain necessary to update it.

I think the 'right' way to fix it might be to provide an azure account for the agent accounts that run the builds so that this this succeeds:

```
$AZTokenRequest = Get-AzAccessToken -ResourceType Arm
$headers = @{ Authorization = "Bearer $($AZTokenRequest.Token)" }
 ```
 
 Though I have no idea how this SYSTEM_ACCESSTOKEN is set downstream - it's used pretty broadly.